storm-core/src/jvm/backtype/storm/spout/CheckpointSpout.java

```diff
@@ -0,0 +1,280 @@
+        return conf;
+    }
+
+    public static boolean isCheckpoint(Tuple input) {
+        return CHECKPOINT_STREAM_ID.equals(input.getSourceStreamId());
+    }
+
+    /**
+     * Loads the last saved checkpoint state the from persistent storage.
+     */
+    private KeyValueState<String, CheckPointState> loadCheckpointState(Map conf, TopologyContext ctx) {
+        String namespace = ctx.getThisComponentId() + "-" + ctx.getThisTaskId();
+        KeyValueState<String, CheckPointState> state =
+                (KeyValueState<String, CheckPointState>) StateFactory.getState(namespace, conf, ctx);
+        if (state.get(TX_STATE_KEY) == null) {
+            CheckPointState txState = new CheckPointState(-1, COMMITTED);
+            state.put(TX_STATE_KEY, txState);
+            state.commit();
+            LOG.debug("Initialized checkpoint spout state with txState {}", txState);
+        } else {
+            LOG.debug("Got checkpoint spout state {}", state.get(TX_STATE_KEY));
+        }
+        return state;
+    }
+
+    private int loadCheckpointInterval(Map stormConf) {
+        int interval;
+        if (stormConf.containsKey(Config.TOPOLOGY_STATE_CHECKPOINT_INTERVAL)) {
+            interval = ((Number) stormConf.get(Config.TOPOLOGY_STATE_CHECKPOINT_INTERVAL)).intValue();
+        } else {
+            interval = DEFAULT_CHECKPOINT_INTERVAL;
+        }
+        return interval;
+    }
+
+    private boolean shouldRecover() {
+        return recovering && !recoveryStepInProgress;
+    }
+
+    private boolean shouldCheckpoint() {
+        return !recovering && !checkpointStepInProgress
+                && (System.currentTimeMillis() - lastCheckpointTs) > checkpointInterval;
+    }
+
+    private boolean shouldRollback(CheckPointState txState) {
+        return txState.state == PREPARING;
+    }
+
+    private boolean shouldCommit(CheckPointState txState) {
+        return txState.state == COMMITTING;
+    }
+
+    private boolean shouldInitState(CheckPointState txState) {
+        return txState.state == COMMITTED;
+    }
+
+    private void handleRecovery() {
+        CheckPointState txState = getTxState();
+        LOG.debug("Current txState is {}", txState);
+        if (shouldRollback(txState)) {
+            LOG.debug("Emitting rollback with txid {}", txState.txid);
+            collector.emit(CHECKPOINT_STREAM_ID, new Values(txState.txid, CHECKPOINT_ACTION_ROLLBACK), txState.txid);
+        } else if (shouldCommit(txState)) {
+            LOG.debug("Emitting commit with txid {}", txState.txid);
+            collector.emit(CHECKPOINT_STREAM_ID, new Values(txState.txid, CHECKPOINT_ACTION_COMMIT), txState.txid);
+        } else if (shouldInitState(txState)) {
+            LOG.debug("Emitting init state with txid {}", txState.txid);
+            collector.emit(CHECKPOINT_STREAM_ID, new Values(txState.txid, CHECKPOINT_ACTION_INITSTATE), txState.txid);
+        }
+        startProgress();
+    }
+
+    private void handleRecoveryAck() {
```

Comment:
the state transition code seems to be repeating in handleCheckpointAck, doCheckpoint and handleRecoveryAck. Do you think we can capture the nextState and prevState for each state in the CheckPointState class it self so in all these methods we can just call netxt/prevState followed by saveTxState(txState).